includes/parser/Parser.php

```diff
@@ -1274,12 +1274,26 @@ function internalParse( $text, $isMain = true, $frame = false ) {
 			// XW-4380: Make template placeholders in list items render correctly
 			// XW-4609: remove newlines from template's placeholder when used inside list's item to not break list
 			// before this code was executed inside doBlockLevels(), however it needs to be done before doAllQuotes to not break bolds and italics
-			$text = preg_replace_callback('/^([\*#;:][^\n]*<div class="placeholder placeholder-double-brackets"[^>]+>&#x0200B;)\n?(.*?)(&#x0200B;<\/div>)/ms', function($matches) {
-				return preg_replace('/\\n/', '', $matches[0]);
-			}, str_replace("\r\n", "\n", $text));
-			$text = preg_replace_callback('/^([\*#;:][^\n]*<span class="placeholder placeholder-double-brackets"[^>]+>&#x0200B;)\n?(.*?)(&#x0200B;<\/span>)/ms', function($matches) {
-				return preg_replace('/\\n/', '', $matches[0]);
-			}, str_replace("\r\n", "\n", $text));
+
+			do {
+				$before = $text;
+				$text = preg_replace_callback('/^([\*#;:][^\n]*<div class="placeholder placeholder-double-brackets"[^>]+>&#x0200B;)(.*?)(&#x0200B;<\/div>)/ms',
```

Comment:
Could we run these checks inside preprocessor_dom, straight after a decision about which placeholder (div or span) should be displayed? This way we could lower the complexity of these regexes :)

Reply:
@bart unfortunately not, in preprocessor_dom we have access only to template's wikitext so we don't know if it is used inside list or not. Here we have access to whole page wikitext and these regexes checks if given line starts with *, #, ; or : which are used to create list

Reply:
OK, I get it.